sources/HUBViewModelDiff.m

```diff
@@ -145,15 +163,234 @@ + (instancetype)diffFromViewModel:(id<HUBViewModel>)fromViewModel
                                              reloads:[reloads copy]];
 }
 
-- (NSString *)debugDescription
+typedef struct {
+    NSInteger x;
+    NSInteger y;
+} HUBDiffPoint;
+
+static inline HUBDiffPoint HUBDiffPointMake(NSInteger x, NSInteger y) {
+    return (HUBDiffPoint){ .x = x, .y = y};
+}
+
+typedef struct {
+    NSInteger D;
+    NSInteger k;
+    NSInteger previousX;
+    NSInteger nextX;
+} HUBDiffTraceStep;
+
+static inline HUBDiffTraceStep HUBDiffTraceStepMake(NSInteger D, NSInteger k, NSInteger previousX, NSInteger nextX) {
+    return (HUBDiffTraceStep){ .D = D, .k = k, .previousX = previousX, .nextX = nextX };
+}
+
+typedef NS_ENUM(NSUInteger, HUBDiffTraceType) {
+    HUBDiffTraceTypeInsert,
+    HUBDiffTraceTypeDelete,
+    HUBDiffTraceTypeMatchPoint
+};
+
+@interface HUBDiffTrace : NSObject
+
+@property (nonatomic, assign, readonly) HUBDiffPoint from;
+@property (nonatomic, assign, readonly) HUBDiffPoint to;
+@property (nonatomic, assign, readonly) NSInteger D;
+
+@property (nonatomic, assign, readonly) HUBDiffTraceType type;
+
+@end
+
+@implementation HUBDiffTrace
+
++ (instancetype)nextTraceFromStep:(HUBDiffTraceStep)step
 {
-    return [NSString stringWithFormat:@"\t{\n\
-        deletions: %@\n\
-        insertions: %@\n\
-        reloads: %@\n\
-    \t}", self.deletedBodyComponentIndexPaths, self.insertedBodyComponentIndexPaths, self.reloadedBodyComponentIndexPaths];
+    HUBDiffTraceType type;
+    if (step.k == -(step.D)) {
+        type = HUBDiffTraceTypeInsert;
+    } else if (step.k != step.D) {
+        if (step.previousX < step.nextX) {
+            type = HUBDiffTraceTypeInsert;
+        } else {
+            type = HUBDiffTraceTypeDelete;
+        }
+    } else {
+        type = HUBDiffTraceTypeDelete;
+    }
+    
+    if (type == HUBDiffTraceTypeInsert) {
+        NSInteger x = step.nextX;
+        
+        return [[self alloc] initWithFromPoint:HUBDiffPointMake(x, x - step.k - 1) toPoint:HUBDiffPointMake(x, x - step.k) D:step.D];
+    } else {
+        NSInteger x = step.previousX + 1;
+
+        return [[self alloc] initWithFromPoint:HUBDiffPointMake(x - 1, x - step.k) toPoint:HUBDiffPointMake(x, x - step.k) D:step.D];
+    }
+}
+
+- (instancetype)initWithFromPoint:(HUBDiffPoint)fromPoint toPoint:(HUBDiffPoint)toPoint D:(NSInteger)D
```

Comment:
How about naming the `D` parameter `minimumNumberOfChanges` instead? Since that’s what it represents if I understand your API docs further up correctly? Would make the code a bit more readable 😄